bottlerocket-os/bottlerocket
# netdog: Support versioning in `net.toml`

**Issue number:**
Implements #2214 


**Description of changes:**
```
This change adds the ability to handle additional versions of `net.toml`
as the project continues to add features to network configuration.  A
top level `NetConfig` type has been removed in favor of a few traits.
Each new version of config is expected to implement these traits, the
most important of which converts the network configuration into
`WickedInterface` structs which are suitable for serializing directly to
file.

As part of this change, the `net_config` module has been lightly
refactored and split into submodules to make the pieces easier to use
among new versions.
```


**Testing done:**
* All unit tests continue to pass
* Build and boot an `aws-k8s-1.21 variant` and ensure the network config is correct
* Build and boot a `metal-dev` variant and ensure the network config is correct via a given `net.toml`


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
